classes/converter.php

```diff
@@ -53,20 +74,29 @@ abstract class converter {
      * @param string $key access key to use for user validation, this is required to login user and allow access of target page
      * for conversion {@see \tool_pdfpages\helper::create_user_key}.
      * @param string $filename the name to give converted file.
-     * (if none is specified, use {@see \tool_pdfpages\helper::get_moodle_url_pdf_filename})
+     * (if none is specified, filename will be generated {@see \tool_pdfpages\helper::get_moodle_url_pdf_filename})
      * @param array $options any additional options to pass to converter, valid options vary with converter
      * instance, see relevant converter for further details.
      * @param string $cookiename cookie name to apply to conversion (optional).
      * @param string $cookievalue cookie value to apply to conversion (optional).
      *
      * @return \stored_file the stored file created during conversion.
      */
-    public function convert_moodle_url_to_pdf(moodle_url $url, string $key, string $filename = '', array $options = [],
+    final public function convert_moodle_url_to_pdf(moodle_url $url, string $key, string $filename = '', array $options = [],
                                               string $cookiename = '', string $cookievalue = ''): \stored_file {
-        // Implement converter specific logic for URL PDF extraction here.
-        // When implemented, the target URL must be passed through the proxy page (index.php) as a parameter
-        // along with the access key, in order to validate user login.
-        // {@see \tool_pdfpages\helper::get_proxy_url}.
+
+        try {
+            $proxyurl = helper::get_proxy_url($url, $key);
+            $filename = ($filename === '') ? helper::get_moodle_url_pdf_filename($url) : $filename;
+            $content = $this->generate_pdf_content($proxyurl, $filename, $options, $cookiename, $cookievalue);
+
+            return $this->create_pdf_file($content, $filename);
+        } catch (\Exception $exception) {
+            throw new \moodle_exception('error:urltopdf', 'tool_pdfpages', '', null, $exception->getMessage());
+        } finally {
+            // Make sure the access key token session cannot be used for any other requests, prevent session hijacking.
+            \core\session\manager::terminate_current();
```

Comment:
I'm still not convince that we will reach it in all scenarios. Is there any way to add tests? Behat tests?

Reply:
There can be no behat tests as this is happening in a separate process, finally is ALWAYS carried out as per PHP https://www.php.net/manual/en/language.exceptions.php `All finally blocks it encounters along the way will be executed.`

Reply:
can we prove that in tests?

Reply:
another thing to test is that we still can access returned file after the session is terminated. Will we be able to read the content of the file?

Reply:
I have added session termination test for the abstract converter class